libwayshot/examples/waymirror-egl/src/state.rs

```diff
@@ -0,0 +1,305 @@
+
+        let attributes = [
+            egl::SURFACE_TYPE,
+            egl::WINDOW_BIT,
+            egl::RENDERABLE_TYPE,
+            egl::OPENGL_ES2_BIT,
+            egl::RED_SIZE,
+            8,
+            egl::GREEN_SIZE,
+            8,
+            egl::BLUE_SIZE,
+            8,
+            egl::NONE,
+        ];
+
+        let config = self
+            .egl
+            .choose_first_config(self.egl_display.unwrap(), &attributes)?
+            .expect("unable to find an appropriate EGL configuration");
+        self.egl_surface = Some(unsafe {
+            self.egl.create_window_surface(
+                self.egl_display.unwrap(),
+                config,
+                self.egl_window.clone().unwrap().ptr() as egl::NativeWindowType,
+                None,
+            )?
+        });
+
+        let context_attributes = [egl::CONTEXT_CLIENT_VERSION, 2, egl::NONE];
+        self.egl_context = Some(self.egl.create_context(
+            self.egl_display.unwrap(),
+            config,
+            None,
+            &context_attributes,
+        )?);
+
+        self.egl.make_current(
+            self.egl_display.unwrap(),
+            self.egl_surface,
+            self.egl_surface,
+            self.egl_context,
+        )?;
+
+        self.init_program()?;
+
+        Ok(())
+    }
+
+    fn init_program(&mut self) -> Result<()> {
+        let vert_shader = load_shader(
+            gl::VERTEX_SHADER,
+            include_str!("./shaders/vert.glsl").into(),
+        )
+        .unwrap();
+
+        let frag_shader = load_shader(
+            gl::FRAGMENT_SHADER,
+            include_str!("./shaders/frag.glsl").into(),
+        )
+        .unwrap();
+
+        unsafe {
+            self.gl_program = gl::CreateProgram();
+        }
+
+        if self.gl_program == 0 {
+            tracing::event!(tracing::Level::ERROR, "glCreateProgramFailed!");
+            return Err(WaylandEGLStateError::GLCreateProgramFailed);
+        }
+
+        unsafe {
+            gl::AttachShader(self.gl_program, vert_shader);
+            gl::AttachShader(self.gl_program, frag_shader);
+
+            gl::LinkProgram(self.gl_program);
+        }
+
+        let mut linked: gl::types::GLint = 1;
+        unsafe { gl::GetProgramiv(self.gl_program, gl::LINK_STATUS, &mut linked as *mut i32) }
+
+        if linked > 0 {
+            tracing::event!(tracing::Level::INFO, "Successfully linked the program!");
+        } else {
+            return Err(WaylandEGLStateError::GLLinkProgramFailed);
+        }
+
+        let vertices: [gl::types::GLfloat; 20] = [
+            // positions             // texture coords
+            1.0, 1.0, 0.0, 1.0, 0.0, // top right
+            1.0, -1.0, 0.0, 1.0, 1.0, // bottom right
+            -1.0, -1.0, 0.0, 0.0, 1.0, // bottom left
+            -1.0, 1.0, 0.0, 0.0, 0.0, // top left
+        ];
+        let indices: [gl::types::GLint; 6] = [
+            0, 1, 3, // first Triangle
```

Comment:
Why are we creating triangle indices in this example?

Reply:
We're rendering two triangles to form the Rectangle here. The indices index into the vertices array.